### PR TITLE
Introduce publishing

### PR DESCRIPTION
**Pull request by Mitko-Kerezov**
* Introduce publish command
* Introduce a `CloudOptionsProvider` used for managing additional `--` flags 
* Introduce publish service
* Extract some interfaces and cleanup some code

**NB!** Docs and tests are pending

Merge after https://github.com/NativeScript/nativescript-cli/pull/3128

Ping @rosen-vladimirov @TsvetanMilanov

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -35,10 +35,10 @@
     "grunt-shell": "1.3.0",
     "grunt-ts": "6.0.0-beta.16",
     "istanbul": "0.4.5",
-    "mobile-cli-lib": "https://github.com/telerik/mobile-cli-lib/tarball/master",
```
[the remark this answers is not in the document]
Comment (Mitko-Kerezov) [reply package.json]: This is needed in order for the transpilation execute successfully. Will be reverted once I merge the PRs in mobile-cli-lib and nativescript-cli

File: package.json [hunk 2]
```diff
@@ -35,10 +35,10 @@
...
+    "mobile-cli-lib": "https://github.com/telerik/mobile-cli-lib/tarball/kerezov/commands--flags",
     "mocha": "3.5.3",
     "mocha-typescript": "1.1.8",
-    "nativescript": "https://github.com/NativeScript/nativescript-cli/tarball/master",
```
[the remark this answers is not in the document]
Comment (Mitko-Kerezov) [reply package.json]: This is also needed in order for the transpilation execute successfully. Will be reverted once I merge the PRs in mobile-cli-lib and nativescript-cli

File: lib/services/upload-service.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import * as uuid from "uuid";
+
+import { BUILD_SERVICE_NAME, HTTP_METHODS } from "../constants";
+
+export class UploadService implements IUploadService {
+
+	protected serviceName = BUILD_SERVICE_NAME;
+
+	constructor(private $buildCloudService: IBuildCloudService,
+		protected $cloudRequestService: ICloudRequestService,
```
Comment (TsvetanMilanov) [inline lib/services/upload-service.ts]: We don't need the `$cloudRequestService`.

Comment (Mitko-Kerezov) [reply lib/services/upload-service.ts]: 🎣

**Pull request merged**